ext/pdo/pdo_dbh.c

```diff
@@ -221,10 +221,62 @@ static char *dsn_from_uri(char *uri, char *buf, size_t buflen) /* {{{ */
 }
 /* }}} */
 
-/* {{{ */
-PHP_METHOD(PDO, __construct)
+static bool create_driver_specific_pdo_object(pdo_driver_t *driver, zend_class_entry *called_scope, zval *new_object)
+{
+	zend_class_entry *ce;
+	zend_class_entry *ce_based_on_driver_name = NULL, *ce_based_on_called_object = NULL;
+
+	ce_based_on_driver_name = zend_hash_str_find_ptr(&pdo_driver_specific_ce_hash, driver->driver_name, driver->driver_name_len);
+
+	ZEND_HASH_MAP_FOREACH_PTR(&pdo_driver_specific_ce_hash, ce) {
+		if (called_scope != pdo_dbh_ce && instanceof_function(called_scope, ce)) {
+			ce_based_on_called_object = called_scope;
+			break;
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	if (ce_based_on_called_object) {
+		if (ce_based_on_driver_name) {
+			if (instanceof_function(ce_based_on_called_object, ce_based_on_driver_name) == false) {
+				zend_throw_exception_ex(php_pdo_get_exception(), 0,
+					"%s::connect() cannot be called when connecting to the \"%s\" driver, "
+					"either %s::connect() or PDO::connect() must be called instead",
+					ZSTR_VAL(called_scope->name), driver->driver_name, ZSTR_VAL(ce_based_on_driver_name->name));
+				return false;
+			}
+
+			object_init_ex(new_object, ce_based_on_called_object);
+			return true;
+		} else {
+			zend_throw_exception_ex(php_pdo_get_exception(), 0,
+				"%s::connect() cannot be called when connecting to an unknown driver, "
+				"PDO::connect() must be called instead",
+				ZSTR_VAL(called_scope->name));
+			return false;
+		}
+	}
+
+	if (ce_based_on_driver_name) {
+		if (called_scope != pdo_dbh_ce) {
+			zend_throw_exception_ex(php_pdo_get_exception(), 0,
+				"%s::connect() cannot be called when connecting to the \"%s\" driver, "
+				"either %s::connect() or PDO::connect() must be called instead",
+				ZSTR_VAL(called_scope->name), driver->driver_name, ZSTR_VAL(ce_based_on_driver_name->name));
+			return false;
+		}
```

Comment:
Wait does this mean, that one effectively cannot subclass PdoSQLite and co?

Reply:
No it isn't. :) AFAIR I even added tests where I subclass some of the driver specific subclasses. On the other hand, this condition is used to make sure that you cannot instantiate a driver-specific subclass when you use a subclass of `PDO` which is not a subclass of the driver-specific subclass (ah sorry, this sentence sounds like a mindfck). A better example:

```
class MyPdo extends PDO {}

MyPdo::connect('sqlite::memory:');
```

Doing so would contradict to the `static` return type that's why I had to remove support for this case.

P.s. I know this piece of code is very convoluted, but I wasn't able to simply it even more.. :(

Reply:
Ahhhhhhh okay, maybe add some comments to explain this. Because this is not obvious at the beginning

Reply:
I clarified the behavior with some comments